### PR TITLE
Initial classes

### PR DESCRIPTION
**Pull request by mecha**

### CODE REVIEW

File: src/ChildrenAwareRecursiveIteratorTrait.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Common functionality for recursive iterators that iterate over children-aware items.
+ *
+ * @since [*next-version*]
+ */
+trait ChildrenAwareRecursiveIteratorTrait
+{
+    /**
+     *{@inheritdoc}
```
Comment (XedinUnknown) [inline src/ChildrenAwareRecursiveIteratorTrait.php]: Space missing.

File: src/ModeAwareTrait.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * Functionality for something that is aware of a recursion mode.
+ *
+ * @since [*next-version*]
+ */
+trait ModeAwareTrait
+{
+    /**
+     * The recursion mode.
+     *
+     * @see RecursiveIteratorInterface::MODE_SELF_FIRST
```
Comment (XedinUnknown) [inline src/ModeAwareTrait.php]: I imagine that other iterators can have other modes, so perhaps listing recursive iterator modes here can be misleading.

File: src/ModeAwareTrait.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace Dhii\Iterator;
+
+/**
+ * Functionality for something that is aware of a recursion mode.
```
Comment (XedinUnknown) [inline src/ModeAwareTrait.php]: Here and everywhere else in this trait, perhaps write "iteration mode", since the trait does not really imply that the mode is used only for recursion.

File: src/RecursiveIteration.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function getDepth()
+    {
+        return count($this->getPathSegments()) - 1;
```
Comment (XedinUnknown) [inline src/RecursiveIteration.php:71]: Since the path segments indicate the nodes, into which to enter from another node (perhaps root), the depth should IMO be equal to the count of path segments: the root node is not listed as a segment.

File: test/functional/AbstractBaseIterationTest.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace Dhii\Iterator\FuncTest;
+
+use Dhii\Iterator\AbstractBaseIteration;
+use Xpmock\TestCase;
+
+/**
+ * Tests {@see Dhii\Iterator\AbstractBaseIteration}.
```
Comment (XedinUnknown) [inline test/functional/AbstractBaseIterationTest.php]: Root FQN.

**Review by XedinUnknown — CHANGES_REQUESTED** [review]

File: src/RecursiveIteration.php (hunk 4 above)
Comment (mecha) [reply src/RecursiveIteration.php:71]: If I recall, we had found (and also agreed) that depth is the number of "edges" needed to visit the root node.

[Article reference](http://typeocaml.com/2014/11/26/height-depth-and-level-of-a-tree/).

[`DepthAwareInterface`](https://github.com/Dhii/hierarchy-interface/blob/b026362a9b7dc7e8e9473c96239b881c86536040/src/DepthAwareInterface.php#L17-L18)

Comment (XedinUnknown) [reply src/RecursiveIteration.php:71]: Exactly @mecha . If you have a path of 3 segments, each segment is a name of a node to enter. Thus there are 3 nodes, plus root. In [this tree](http://typeocaml.com/content/images/2014/11/edge-1.jpg#small), you need 3 segments to consecutively enter nodes the times in order to get from root to "E": C -> D -> E. That's a path of 3. Depth is 3. No?

Comment (mecha) [reply src/RecursiveIteration.php:71]: Yes, in that example the depth would be 3, but the path wouldn't be 3.  
The path would have to be `A -> C -> D -> E`; that's 4 segments.

Comment (XedinUnknown) [reply src/RecursiveIteration.php:71]: The path is never `A -> C -> D -> E`. The path is from the node where you are now. For absolute paths, that's the root node. You are in `A`. It's `C -> D -> E`.

Comment (mecha) [reply src/RecursiveIteration.php:71]: Disregard my previous comment. I got confused.

A path of 3 segments has a depth of 2. When calculating depth, the baseline is the root node of the tree. Given that the root node has a depth of zero, but the root can still have a key and a path segment, it follows that the depth is the number of path segments minus 1.

Once again, I will reference [the article](http://typeocaml.com/2014/11/26/height-depth-and-level-of-a-tree/) that outlines the concepts we agreed on.

Comment (XedinUnknown) [reply src/RecursiveIteration.php:71]: @mecha, I really find it hard to follow the logic you described, because to me the logic seems sound, but the conclusion seems wrong.

In the example I gave, where the path is `C -> D -> E`, you are in node `A`, and therefore, you have the following transitions:

1. `A -> C`
2. `C -> D`
3. `D -> E`

That's a total of 3 transitions. For a path of 3 segments. 3 = 3. Number of transitions equals the number of segments.

Comment (XedinUnknown) [reply src/RecursiveIteration.php:71]: Look at the explanation of Depth in the page you linked to. The depth of `D` is 2. Why? Because the path is `C/D`, i.e. `A -> C -> D`. Transitions `A -> C` and `C -> D`. 2 transitions. 2 segments. Depth of 2.

Comment (mecha) [reply src/RecursiveIteration.php:71]: I think we're agreeing on how depth is defined but disagreeing on paths. You said:

> the path is C/D, i.e. A -> C -> D.

To me, the path segments are `[A, C, D]`, hence the path is made up of 3 segments and therefore `depth = 3 - 1 = 2`.

I cannot make sense of a path like `C/D`; indeed if this is the kind of thing that you consider a path, then the number of segments would equal the depth. But the root is a node just like any other except it has no parent, so I believe it should be a part of the path.

Comment (XedinUnknown) [reply src/RecursiveIteration.php:71]: The root, however, is not something you enter, just like any other node in which you currently are. The path describes a series of nodes *from your current node*. Otherwise, you are describing something like a segment, which is arbitrary. But in your case, it is not arbitrary; it is from *the current node*.

Comment (mecha) [reply src/RecursiveIteration.php:71]: I think you may be right, if you only look at how we achieve iteration and what the path segments signify in that regard. Conceptually, I think it is a problem that we cannot solve **in the scope of iteration**. I'll explain:

In the scope of the interface for which we are having this discussion, `RecursiveIterationInterface`, the path segments are a property of the iteration, which in turn describe the path to the element that was iterated over (in other words, will be yielded at some point by `current()`).

---

Iteration will always be initiated on something, be it the first node in a hierarchy of objects or the outside-most array, to give an example.

```
[
    "a": ["b", "c", "d"],
    "e": ["f"]
]
```

In a case such as the above, the "root" can be considered to be the array construct, which has two immediate children. It's key is not known to us. In a recursive algorithm, the value of `a` would eventually be considered a root itself with three elements.

From this point of view, element `b` has a path of `[b]`, and since it is not the root, it has depth of one (equal to path segments). Once you start backtracking, the key of the parent becomes known (`a`) and the path of its children elements must now be relative to the new root (becoming `[a, b]` with a dept of 1). This is what we're doing in `AbstractRecursiveIterator`.

Given that the topmost root key may never be known, and may even not exist, the root cannot be considered a node in this regard. So you're right that when iterating, the root is not something that you enter into. 🆗 

----

Similarly, in an hierarchy made up of objects that hold some value:

```
Node [R] : 5
    Node [A] : 12
        Node [B] : 20
        Node [C] : 28
        Node [D] : 24
    Node [E] : 15
        Node [F] : 29
```

In **traversal**, Node `R` will be visited first. This is a case when the key of the root can be known. When **iterating** over `R` however, you lose this information because the iterator will only acknowledge the children.

Once again, consider if the above object hierarchy is but a sub-hierarchy of a larger one. Discarding the root from the path of an element causes you to lose information. If the path of `C` is just `[A, C]`, then when you take the whole super hierarchy into consideration, the actual path is broken.

However note, that this only happens when iterating, not traversing, and can also be avoided. In fact, our abstract iterator implementation does this. When iterating recursively, the path will continue to be "compiled" during backtracking until the only missing path segment is the top-most root. 🆗 

---

These demonstrate the point that you are trying to make. Iterating over a node means iterating over its children, which results in the node being iterated over not being visited. This leads to paths missing the root node, which would make depth equal to the number of path segments, as you said. 🆗 

What I mean is that doing `foreach ($node as $child)` does not yield `$node` as a `$child` at any point in the iteration.

Root nodes are treated in the same way are regular nodes in computer science, for consistency. But perhaps it is wrong to try and apply **traversal** concepts to **iteration**.

In any case, I will change the code such that `depth = number of path segments`. 🎉

Comment (XedinUnknown) [reply src/RecursiveIteration.php:71]: Yep. For me it's as simple as thinking of a path `A/C` as: "go to A", "go to C". Every "go to" is a segment. 2 segments = depth of 2. Height of 3 ;)

Please proceed.

**Review by XedinUnknown — APPROVED** [review]
